toc/rfc/rfc-draft-v2-app-manifests.md

```diff
@@ -0,0 +1,642 @@
+manifest schema and behavior described by this RFC.
+
+#### Fully Declarative (Pruning)
+
+Version 1 app manifests are currently inconsistently-declarative. For instance,
+removing a buildpack from the `buildpacks` sequence will delete the buildpack
+from the app configuration, however removing a route from the `routes` sequence
+will NOT unmap the route from the app.
+
+Version 2 app manifests will be fully-declarative; applying a manifest will
+update state in the space to exactly match the manifest, including deleting
+resources and destructively updating configuration.
+
+Declaratively applying app manifests will have multiple benefits,
+including:
+1. Internally-consistent behavior between manifest nodes
+1. Easier onboarding and reduced confusion for users, since declarative
+   configuration is industry-standard (e.g. BOSH, Concourse, Kubernetes)
+1. Enables more powerful workflows for Cloud Foundry app developers (e.g. GitOps)
+
+Declarative app manifests will be a major change to how app manifests are
+applied and will be a barrier to adoption. It is worth evaluating strategies to
+ease the migration, for instance adding support for "dry runs" and rich
+[manifest
+diffs](https://v3-apidocs.cloudfoundry.org/version/release-candidate/#generate-a-manifest-for-an-app).
+
+It will also be important to be vigilant about breaking changes going forward;
+declarative manifests can exacerbate the impact of accidental breaking changes.
+Even changes as innocuous as adding a manifest node for an existing API field
+can be a breaking change, since existing manifests will not have the
+corresponding node present.
+
+##### Prior Art
+
+Since this is a major behavioral change for app manifests, here a survey of
+some prior art for declarative configuration for similar products:
+- https://blog.nelhage.com/post/declarative-configuration-management/
+- https://argo-cd.readthedocs.io/en/stable/user-guide/auto_sync/#automatic-pruning
+- https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#alternative-kubectl-apply-f-directory-prune
+- https://dzx.fr/blog/declarative-kubernetes-resource-management/
+- https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/3659-kubectl-apply-prune
+
+#### Not Transactional
+
+Like v1 manifests, applying a v2 manifest will NOT be transactional. The
+configuration specified in the manifest will be applied in a non-deterministic
+order and can fail when partially applied. Changes resulting from applying the
+manifest will NOT be rolled-back in the event of a failure.
+
+### Scope of Work
+
+#### API
+
+1. Add a `GET /v3/space/:guid/manifest` endpoint with a `version` query
+   parameter for generating v2 app manifests
+1. Update `POST /v3/spaces/:guid/actions/apply_manifest` to accept v2 manifests
+1. Update `POST /v3/spaces/:guid/manifest_diff` to accept v2 manifests
+
+#### CLI
+
+1. Update `cf push` and `cf apply-manifest` to accept v2 manifests
```

Comment:
**thought** Maybe too detailed for this proposal, but it seems the existing `cf apply-manifest` does display a diff but does theres [no way](https://github.com/cloudfoundry/cli/blob/main/command/v7/apply_manifest_command.go#L20-L25) to wait for user confirmation to apply changes - it just applies changes. In this manifest v2 world, if its possible some actions are destructive we will definitely need that.

Reply:
That may be considered a breaking change (imagine a continuous delivery pipeline that starts hanging/failing due to interactive prompts). We may need to reserve the prompt behavior (or v2 manifest support in general) for a new major version of the CLI.